### PR TITLE
Build/Publish base on GitHub Actions

### PR DESCRIPTION
**Pull request by tomaszgolebiowski**
CLOSE https://linear.app/sourcegraph/issue/CODY-4276/improve-workflow-to-build-vs-based-on-cody-commit

- Migration from build.cake to GitHub Actions.
- Release automation (simply add the appropriate tag with the version number ex. preview-v0.1.3)
- Automatic setting of version in source code files 
- Preview version available as a private feed (https://sourcegraph.github.io/cody-vs/feed.xml)
- Production version automatically published to VS Marketplace
- Nightly builds automatically fetch the latest tagged version of the agent and use it for testing. 
- The script for building the agent now uses PowerShell.

### CODE REVIEW

File: .github/workflows/publish-preview.yml [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+      #Publish
+          
+      - name: Create GitHub release
+        uses: ncipollo/release-action@v1.14.0
+        with:
+          tag: ${{ steps.version.outputs.full }}
+          name: Cody for Visual Studio ${{ steps.version.outputs.full_without_prefix }}
+          prerelease: true
+          artifacts: src/Cody.VisualStudio/bin/${{ env.Configuration }}/Cody.VisualStudio.vsix
+          artifactContentType: application/zip
```
Comment (PiotrKarczmarz) [inline .github/workflows/publish-preview.yml:117]: Could we turn off zipping `.vsix` installer?

**Review by PiotrKarczmarz — COMMENTED** [review]
Let's remove `compiled: ...` from the first line of the log (`Cody` output window), because it doesn't return valid results anymore.

image

File: agent/buildAgent.ps1 [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+$nodeBinFile = Join-Path $nodeDir "node-win-x64.exe"
+$nodeArmBinFile = Join-Path $nodeDir "node-win-arm64.exe"
+$nodeUrl = "https://github.com/sourcegraph/node-binaries/raw/main/v20.12.2/node-win-x64.exe"
+$nodeArmUrl = "https://github.com/sourcegraph/node-binaries/raw/main/v20.12.2/node-win-arm64.exe"
+
+$exclude = @(
+    "src",
+    "scripts",
+    "*.map",
```
Comment (PiotrKarczmarz) [inline agent/buildAgent.ps1:24]: Nice touch by removing *.map files, including index.js.map > 30 MB!

File: .github/workflows/publish-preview.yml (hunk 1 above)
Comment (tomaszgolebiowski) [reply .github/workflows/publish-preview.yml:117]: The installer is not zipped. It is just an additional piece of information that the blob of bytes being transferred is a zip archive.

Comment (PiotrKarczmarz) [reply .github/workflows/publish-preview.yml:117]: Could be this turned off? It's pretty annoying during testing that it's is not clean `.visx` file.

**Comment by PiotrKarczmarz** [comment]
The preview release feed works nicely! 

image

and after switching to PowerShell scripts there is a much more verbose output, when running `runBuildAgent.ps1`. Especially for git step:

image

Also, running `runBuildAgent.ps1` is about 4 seconds faster than `dotnet cake` 🚀

File: .github/workflows/publish-preview.yml (hunk 1 above)
Comment (PiotrKarczmarz) [reply .github/workflows/publish-preview.yml]: Looks like it's not supported https://github.com/actions/upload-artifact/issues/426 😥

**Review by PiotrKarczmarz — APPROVED** [review]

**Pull request merged**